src/fm.py

```diff
@@ -58,6 +58,30 @@ def sa_construct(x: str) -> list[int]:
     return [len(x) - len(i) for i in suf]
 
 
+def fm_search(x: rotating_string, p: str, sa: list[int]) -> list[int]:
```

Comment:
There isn't anything wrong, per se, with writing a function that takes both `x` and `p` as input and then searches, but your application will get one genome (a small set of `x` strings) and a bunch of patterns (a large set of `p` strings), so it makes sense to separate the preprocessing of `x` from the search. That way, you only need to preprocess `x` once and can use the preprocessing data structures for all the patterns after that.

Add to that, that part of the project involves adding an option, `-p`, to the `fm` program for preprocessing the genome. The idea is that you can call `fm -p genome.fa` where you only get the genome and not the reads. This command will preprocess `x` and write it to disk. Then, when you later call `fm genome.fa reads.fq` you can load the preprocessed data, so you don't have to run the preprocessing algorithms again.

One way to move towards this could be to split this function into two. One that takes `x` and create the `C` and `O` tables, and another that take as input the two tables and `p` and does the FM-index search for `p` in `x`. If you have refactored the code that way, you can easily adapt the program so the `-p` option calls the function that generate the tables and then write those to disk (you can use something like `pickle` if you want), and without the `-p` option you can read the tables and call your search function.